### ISSUE TITLE
dependency DAG is too fragile

### ISSUE THREAD

**Issue by sleak-lbl** [spack/spack#5241]
This overlaps a number of other issues around dependencies, and maybe the new concretizer will help.

It seems that any small change anywhere changes the hash of anything remotely dependent on it, and this causes a cascade of issues:

 - updating a package.py file with a new version of a package changes the hash, now nothing will use the previously-installed version. Installing anything new that directly or indirectly uses this package triggers a rebuild of everything in the dependency tree. We've effectively turned an O(n) problem into an O(n^2) one ...

 - If a "lower level" package turns out to have an error, I can't rebuild it without first uninstalling it, which means either forcing the uninstall or uninstalling everything that uses it too. Then when I try to rebuild it, I can't get the original hash (presumably something else has changed somewhere in the system), so I need to uninstall and reinstall everything anyway. But the impact is that a problem in a rarely-used library breaks more important packages that "depend" on it

 - there is also a reproducibility problem, 2 days later I can't recreate a build exactly as it was. I'm not sure what changed in the meantime (maybe because of updates to my packages.yaml file? or to other package.py files?). But "spack spec" shows a different hash, and an attempt to reinstall a package rebuilds the entire tree

 - another ticket has raised the "new install of cmake for every package that uses cmake" - I think this stems from the same underlying issue

The effect is that the system seems very fragile, a minor change can break a large installed base of software. I'm a bit stuck for ideas on how to get around it - maybe getting the concretizer to favor already-installed versions will help?

(edit: fix typo)

**Comment by davydden** [comment]
Fyi https://github.com/LLNL/spack/pull/4595

**Comment by pramodskumbhar** [comment]
@sleak-lbl : this may not actually answer your questions but could be helpful:

*  If someone add new version, Spack will choose latest one. You can avoid this by creating `packages.yaml` with version / variant preferences
* I use spack in development workflow. On my local system I use many packages installed by `brew`, `apt` etc and on hpc system I use pre-installed modules. Spack has excellent support for external packages. This avoid rebuilding lot of stuff.

**Comment by davydden** [comment]
> there is also a reproducibility problem, 2 days later

to be fair, that's because Spack is in alpha and has no stable releases yet.

**Comment by alalazo** [comment]
duplicate of #646

**Comment by alalazo** [comment]
> But the impact is that a problem in a rarely-used library breaks more important packages that "depend" on it

Care to share a reproducible use case, or is it a generic complaint?

**Comment by sleak-lbl** [comment]
the looser concretization and separate build vs link/run concretizations should help (actually, the proposal in #4595 would solve a different problem I have, where two branches of the dependency tree require the same package, but one requires a specific variant)

the reproducibility problem I'm seeing at the moment is with arpack-ng, I build slepc, which built arpack-ng, but for some reason the static libs of arpack-ng were not built (the spec suggested they should have been). So I attempted to rebuild that specific hash of arpack-ng, spack baulked due it being already installed .. so I uninstalled it and tried again. But now I can't get the same hash back, and since the path is used within slepc I'll need to rebuild slepc again. I suspect that there was a "git pull" in the meantime and something changed enough that the hash also changed, so the "still in alpha/beta" comment is valid...

That was the specific case that triggered the comment, but it's also partly triggered by the more generic complaint that every time I build anything, I seem to end up building a new version of every package in the dependency tree, and no manipulation of the written spec seems to result in the hash of something already installed (as per #646)

I'll tinker more with packages.yaml, I guess the workflow would be something like:
 - build something with spack
 - make a modulefile for it
 - edit packages.yaml to view it as an external package for future uses
